presto-parquet/src/main/java/io/prestosql/parquet/predicate/TupleDomainParquetPredicate.java

```diff
@@ -389,4 +502,353 @@ private static ParquetCorruptionException corruptionException(String column, Par
         }
         return Domain.create(ValueSet.all(type), hasNullValue);
     }
+
+    // rangeStatistics is neither null nor empty, checked by the caller
+    private static <F, T extends Comparable<T>> Domain createDomain(Type type,
+                                                                    boolean hasNullValue,
+                                                                    List<ParquetRangeStatistics<F>> rangeStatistics,
+                                                                    Function<F, T> function)
+    {
+        Range firstRange = null;
+        Range[] restRanges = new Range[rangeStatistics.size() - 1];
+        for (int i = 0; i < rangeStatistics.size(); i++) {
+            F min = rangeStatistics.get(i).getMin();
+            F max = rangeStatistics.get(i).getMax();
+            Range range;
+            if (min != null && max != null) {
+                range = Range.range(type, function.apply(min), true, function.apply(max), true);
+            }
+            else if (max != null) {
+                range = Range.lessThanOrEqual(type, function.apply(max));
+            }
+            else if (min != null) {
+                range = Range.greaterThanOrEqual(type, function.apply(min));
+            }
+            else {
+                return Domain.create(ValueSet.all(type), hasNullValue);
+            }
+
+            if (i == 0) {
+                firstRange = range;
+            }
+            else {
+                restRanges[i - 1] = range;
+            }
+        }
+
+        return Domain.create(ValueSet.ofRanges(firstRange, restRanges), hasNullValue);
+    }
+
+    private static <T extends Comparable<T>> Domain createDomain(Type type, ColumnIndex columnIndex, ParquetDataSourceId id, boolean hasNullValue, String columnName, List<T> mins, List<T> maxs)
+            throws ParquetCorruptionException
+    {
+        if (mins.size() == 0 || maxs.size() == 0 || mins.size() != maxs.size()) {
+            return Domain.create(ValueSet.all(type), hasNullValue);
+        }
+        int pageCount = columnIndex.getMinValues().size();
+        List<ParquetRangeStatistics<T>> ranges = new ArrayList<>();
+        for (int i = 0; i < pageCount; i++) {
+            T min = mins.get(i);
+            T max = maxs.get(i);
+            if (min.compareTo(max) > 0) {
+                throw corruptionException(columnName, id, columnIndex);
+            }
+            //TODO:
+            if (min instanceof Long) {
+                ParquetIntegerStatistics statistics = new ParquetIntegerStatistics((Long) min, (Long) max);
+                ranges.add((ParquetRangeStatistics<T>) statistics);
+            }
+            else if (min instanceof Double) {
+                ParquetDoubleStatistics statistics = new ParquetDoubleStatistics((Double) min, (Double) max);
+                ranges.add((ParquetRangeStatistics<T>) statistics);
+            }
+            else if (min instanceof Slice) {
```

Comment:
This is a weird structure. The method is generic, but it switches on runtime type. The caller also switches on column type. It would be cleaner to have 3 methods (one for each type), or to abstract this logic via a function passed to this method.

Reply:
Especially because the types aren't primitive, I think the easiest solution is to have a parameter of type `BiFunction<? super T, ? super T, ? extends ParquetRangeStatistics<T>>`.

(But see my next comment.)